File: 111.二叉树的最小深度.cpp
/*
 * @Author: Zhou Hao
 * @Date: 2021-05-07 10:41:10
 * @LastEditors: Zhou Hao
 * @LastEditTime: 2021-05-07 11:29:08
 * @Description: file content
 * @E-mail: 2294776770@qq.com
 */
/*
 * @lc app=leetcode.cn id=111 lang=cpp
 *
 * [111] 二叉树的最小深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include<vector>
#include<string>
#include<iostream>
#include<unordered_map>
#include<unordered_set>
#include<algorithm>
#include<queue>
#include<numeric>


using namespace std;
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (root) {
            int height = 1, cnt = 0;
            queue<TreeNode*> que;       //用队列储存每一层的所有节点
            que.push(root);
            while (!que.empty()) 
            {
                int sz = que.size();
                for (int i=0; i<sz; ++i)        //遍历这一层的元素
                {
                    root = que.front();     //这里不能直接pop
                    if (!root->left && !root->right) return height; //判断叶子节点
                    if (root->left)  que.push(root->left);   //队列中加入下一层的元素
                    if (root->right)  que.push(root->right);
                    que.pop();      //C++队列pop不返回元素
                }
                ++height;
            }
        }
        return 0;
    }
};
// @lc code=end

File: 111.二叉树的最小深度.py
'''
Author: Zhou Hao
Date: 2021-04-08 12:24:12
LastEditors: Zhou Hao
LastEditTime: 2021-05-07 11:13:03
Description: file content
E-mail: 2294776770@qq.com
'''
#
# @lc app=leetcode.cn id=111 lang=python3
#
# [111] 二叉树的最小深度
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    '''BFS'''
    # def minDepth(self, root: TreeNode) -> int:
    #     if not root:
    #         return 0

    #     cur,res = [root],0
    #     while cur:
    #         res +=1
    #         next_lay = []
    #         for node in cur:
    #             if node.left:
    #                 next_lay.append(node.left)
    #             if node.right:
    #                 next_lay.append(node.right)

    #             if not node.left and not node.right:
    #                 return res
    #         cur = next_lay



    '''DFS,preorder'''
    def minDepth(self, root: TreeNode) -> int:
        if not root:return 0
        

        def dfs(root):
            if not root:return 0

            elif not root.left and not root.right:  return 1

            else:
                child_deepth = []
                if root.left:
                    child_deepth.append(dfs(root.left))
                if root.right:
                    child_deepth.append(dfs(root.right))
                return min(child_deepth)+1




        return dfs(root)


# @lc code=end

